### DISCUSSION TITLE
What are you working on next?

### DISCUSSION THREAD

**Discussion by jgarte** [trealla-prolog/trealla-prolog#578] [General]
@infradig Hi, what are you working next for trealla?

Just curious what's on the roadmap.

all the best,

jgart

https://whereis.みんな/

**Comment by infradig** [comment]
The roadmap is...

  - Get CLP(B) working
  - better tail-call optimization (TCO)
  - code-cleanup and improvements

That's about it for now, there are no planned new features.

**Comment by jgarte** [comment]
Thanks for the info!

What is involved in the code-cleanup and improvements?

Is there anything I can do? I have basic C experience (Made linked-lists, basic algorithms, and data structures in C).

**Comment by infradig** [comment]
Most of the code cleanup involved redoing all the ad-hoc string handling to use the safe string-buffer macros. I guess that's mostly done. 

In some place instruction sequences are built up byte-by-byte in a very clumsy and verbose manner, whereas in other later places it got written using helper functions.... but even so there are still pre-computed lengths appearing as constants. When you have nested structs it gets awkward and error prone.

Of course my dream would be for someone to rewrite print.c which, unsurprisingly, handles term writing. It's the oldest code in the system but has been reworked many times and consists of many hacks pertaining to precedence, operator types, parenthesis and spacing. It should be written from first principles with an eye on the ISO spec. This being stand-alone would be easiest for someone to do, but still in my mind daunting.

**Reply by triska** [reply]
It seems desirable to implement term writing in portable&nbsp;Prolog, maybe that could be an interesting starting point for anyone interested?

**Comment by infradig** [comment]
Something else i'd like to see contributions on are FFI modules. There are already SQLite, Raylib & GSL. More possibilities might be Postgres (and other DBs),  SDL etc. As long as there is an installable shared library & well-defined C interface.

**Comment by infradig** [comment]
Well here's a possible task for someone. Make optional the use of libedit (Editline) , aka libedit-dev on Ubuntu, instead of libreadline (GNU Readline). Should just require changes to tpl.c, history.c & Makefile.

**Comment by infradig** [comment]
I would like to see editline available as an alternative to GNU readline. Their interfaces are not compatible but probably feature comparable. Currently isocline is available but seems problematic & is only there for Windows.

**Comment by infradig** [comment]
Another thing i'd like is to copy src/bif_sregex.c to src/bif_regex.c and modify to use POSIX regex (that is <regex.h>).  And I suppose rename the predicates so both sets could be used... timing comparisons would be useful.

**Comment by infradig** [comment]
Should be starting on Trealla version 3 soon. It will focus on a new memory management & garbage collection scheme, designed to address some shortcomings of the current system.

**Reply by jgarte** [reply]
Hi, what's wrong with the current memory management? Just curious. Might be a starting point for me to read some more Trealla code more carefully :)

**Reply by infradig** [reply]
Currently environment frames/slots can't be released out of order (no holes possible) as it uses just one big chunk of memory, and it relies on backtracking to free things up. This ties up a lot of unused memory.

**Reply by jgarte** [reply]
What memory management strategies are you thinking about applying instead in order to allow holes?

By environment frames you mean lexical environment frames or something else? 🦆
